src/dsl/schema/deserialization.rs

```diff
@@ -64,39 +65,49 @@ where
     let yaml_mapping = value
         .as_mapping()
         .ok_or_else(|| Error::custom(format!("schema is not a yaml mapping - {:#?}", value)))?;
-    let mut type_information = deserialize_object_type(&yaml_mapping)?;
-
-    if type_information.is_none() {
-        let raw_type = RawObjectType::Object;
-        let type_data = ObjectTypeData::with_raw_type(raw_type);
-        type_information = Some(vec![ObjectType::Required(type_data)]);
-    }
-
-    let annotations = serde_yaml::from_value(value.clone())
-        .map_err(|e| Error::custom(format!("cannot deserialize schema annotations - {}", e)))?;
+    let type_information = type_information(&yaml_mapping)?;
 
+    let annotations = annotations(value)?;
     let annotations = annotations_from_type(annotations, &type_information);
+    let properties = properties(yaml_mapping)?;
+    let mapping = mapping(yaml_mapping)?;
+    let when = when(yaml_mapping)?;
+    let formula = formula(yaml_mapping)?;
 
-    let properties = yaml_mapping.get(&Value::from("properties"));
-    let properties = match properties {
-        None => None,
-        Some(properties) => match properties {
-            Value::Sequence(sequence) => Some(sequence_to_schema_list(&sequence.to_vec())?),
-            _ => return Err(Error::custom("`properties` is not a yaml sequence")),
-        },
-    };
+    Ok(Schema {
+        types: type_information,
+        children: properties,
+        mapping: mapping.cloned(),
+        annotations,
+        when,
+        formula,
+    })
+}
 
-    let mapping = yaml_mapping.get(&Value::from("mapping"));
-    let mapping = match mapping {
+fn formula<E>(yaml_mapping: &Mapping) -> Result<Option<String>, E>
+where
+    E: Error,
+{
+    let value = yaml_mapping.get(&Value::from("formula"));
+    match value {
         None => Ok(None),
-        Some(mapping) => match mapping {
-            Value::Mapping(mapping) => Ok(Some(mapping)),
-            _ => Err(Error::custom(format!("cannot deserialize mapping {:#?}", mapping))),
+        Some(value) => match value {
+            Value::String(string) => Ok(Some(string.to_string())),
```

Comment:
@cyplo just curious, will it work for following cases?

* `formula: 1` (returns `Value::Number(1)`)
* `formula: true`(returns `Value::Boolean(true)`)

All these are valid formulas. In other words, it doesn't look like it's all coerced to `Value::String` or is it?

Reply:
If not and it needs to be updated, please, add tests for these cases where simple primitive value is a valid formula as well.

Reply:
If it's a string as per `serde_yaml` understanding of a string then this arm will match. If not - see the next arm, which does 'please give me a json string for this yaml value' trick to avoid this situation. I tried with `formula: -7` as it is in the tests. I just tried manually with `true` and seems working as well.

Reply:
Hope this helps, maybe it means that this is unclear in the code or in the tests, so I should change that ? I think the bahviour is correct, the code may look weird though - what do you think @zrzka ?

Reply:
Ok, I just asked as I wasn't sure. And because cdsl future is fuzzy :D let's keep it :)

Reply:
no probs, good catch, this case tripped me as well :)

Reply:
BTW I'm checking these PRs from the "semantic" point of view, not looking at the code in detail. Just FYI.